generators/templates/__macros.j2

```diff
@@ -0,0 +1,53 @@
+{% macro make_import(data) %}
+{%- if data.properties | length > 0 %}
+from {{ data.exercise | to_snake }} import {{ data.properties | map("to_snake") | join(", ") }}
+{%- else %}
+import {{ data.exercise | to_snake }}
+{% endif %}
+{% endmacro %}
+
+{% macro make_tests(data) %}
+class {{ data.exercise | to_camel }}Test(unittest.TestCase):
+    {%- if data.has_error %}
+    {{ add_setup(error=data.has_error) | indent(4) }}
+    {% endif %}
+    {%- for case in data.cases recursive %}
+        {%- if case.cases %}
+        {%- if loop.nextitem is defined %}
+        {{ "#" * loop.depth }} {{ case.description }}
+        {% endif %}
+        {{- loop(case.cases) }}
+        {%- else %}
+        def test_{{ case.description | to_snake }}(self):
+            {%- if case | is_error %}
+            {{- add_assert_raises(case) | indent(12) }}
+            {%- else %}
+            {{- add_assert_equal(case) | indent(12) }}
+            {% endif %}
+        {%- endif %}
+    {%- endfor %}
+{% endmacro %}
+
+{% macro add_setup(error=False) %}
+# Utility functions
+def setUp(self):
+{%- if error %}
+    try:
+        self.assertRaisesRegex
+    except AttributeError:
+        self.assertRaisesRegex = self.assertRaisesRegexp
+{%- endif -%}
+{% endmacro %}
+
+{% macro add_assert_equal(case) %}
+{%- if not case.input %}
+self.assertEqual({{ case.property | to_snake }}(), {{ case | format_expect }})
+{%- else %}
+self.assertEqual({{ case.property | to_snake }}({{ case | format_input }}), {{ case | format_expect }})
+{% endif %}
+{% endmacro %}
+
+{% macro add_assert_raises(case) %}
+self.assertRaisesRegex({{ case.property | to_snake }}({{ case | format_input }}), {{ case | format_expect }})
```

Comment:
By [convention](https://github.com/exercism/python/#conventions), this macro should be:
```
with self.assertRaisesRegex({{ case.error_type | to_camel }}):
	{{ case.property | to_snake }}({{ case | format_input }})
```

Reply:
In most cases, `case.error_type` will be `ValueError`. However, there are some exercises that define their own error types. I suspect some sort of file will have to be created that defines exceptions to the "use ValueError" rule (perhaps yaml would be a good choice for this?).

Reply:
Let me think about that one. Have you got a specific example of an exercise that does this?

Reply:
https://github.com/exercism/python/blob/master/exercises/hamming/hamming_test.py#L53-L55

Reply:
I agree that custom error classes are probably not a requirement for this track and should be removed in the process of creating generators. However, I don't think it is correct (or beneficial to those learning to code for the first time) to always require the same error class or, worse, to allow any error to be thrown.

For example, the Pandas library has [an excellent example](https://github.com/pandas-dev/pandas/wiki/Choosing-Exceptions-to-Raise) of the difference between different types of exceptions, and I think we would be doing students a disservice to instill a habit of throwing any old meaningless exception, instead of teaching the habit of putting thought into your exceptions.

Reply:
I agree, however that might make more sense as an exercise to itself, as the exception hierarchy system in Python is quite dense (and also rather unique) ... like I'd **love** to have something that shows why you *never* `except: do_something()` in production code, but that involves something far more dense than is likely to be able to be defined in  canonical data, much less built via an automated script.

I mean personally I hate what we're doing right now, where we're policing a specific error message (in English), but I cannot see a pathway by which we convert what little we know about the error state being handled in the canonical data to something we could get imported in the template, unless we leave that up to individual, per-exercise template and configuration.

Which I'm not ruling out, it just sounds like a lot of complexity for a relatively small number of known cases.

Reply:
An exercise that required the student to handle some of the _vast_ range of possible OS error symbols defined in [errno](https://docs.python.org/3.7/library/errno.html) would be brilliant, BTW ... and they're relatively standard, which means there _might_ be both some overlap with other languages and a meaningful way of writing canonical properties that would meaningfully map to meaningful tests that could also be meaningfully generated.

Reply:
I just said meaningful way too much ... it's quite late here. I'll come back to this tomorrow.

Reply:
> I mean personally I hate what we're doing right now, where we're policing a specific error message

That's exactly what I'm trying to avoid. All we currently do is ensure that there is indeed a message, and that it is not empty; the Python track does not check for the verbatim error message described in the canonical data.

> Which I'm not ruling out, it just sounds like a lot of complexity for a relatively small number of known cases.

It is; I think it could be done independent of whether the exercise has its own unique template or not. Something like this might be the simplest we could get:

```yml
# generators/errors.yml
forth:
  errors_if_there_is_only_one_value_on_the_stack: IndexError
```

```python
# __macros.j2
{% macro add_assert_raises(case) %}
with self.assertRaisesRegex({{ case | error_type }}):
    {{ case.property | to_snake}}({{ case | format_input }})
{% endmacro %}
```
```python
# generate.py
import yaml
with open('errors.yml') as f:
    errors = yaml.load(f)

def error_type(case):
	case_name = to_snake(case['description'])
	current_exercise = ???
	return errors[current_exercise].get(case_name, 'ValueError')
```

It's still a little messy though, and I agree that it does seem like it might be to much effort for too few use cases.

> An exercise that required the student to handle some of the vast range of possible OS error symbols defined in errno would be brilliant, BTW ... and they're relatively standard, which means there might be both some overlap with other languages and a meaningful way of writing canonical properties that would meaningfully map to meaningful tests that could also be meaningfully generated.

Sounds like a good exercise candidate; you could create an issue in problem-specifications to take the idea further!